src/medicine/medicine.py

```diff
@@ -0,0 +1,70 @@
+class Medicine:
+    """Contains arguments necessary to calculate time after which the medicine will run out."""
+
+    def __init__(self, name: str, dose: int | float, daily_dose: int | float, current_amount: int):
+        """Create a medicine instance.
+
+        :param str name: The name of the medicine
+        :param int|float dose: The dose (in mg) in a single pill
+        :param int|float daily_dose: The dose (in mg) to be taken daily
+        :param int current_amount: the current amount of pills available
+        """
+        if not isinstance(name, str):
+            raise ValueError('name is supposed to be a string')
+        self.name = name
+
+        if not isinstance(dose, (int, float)):
+            raise ValueError('dose is supposed to be a number')
+        if dose < 0:
+            raise ValueError('Negative doses do not exist')
+        self.dose = dose
+
+        if not isinstance(daily_dose, (int, float)):
+            raise ValueError('daily_dose is supposed to be a number')
+        if dose < 0:
+            raise ValueError('Negative doses do not exist')
+        self.daily_dose = daily_dose
+
+        if not isinstance(current_amount, int):
+            raise ValueError('current_amount is supposed to be an integer')
+        if current_amount < 0:
+            raise ValueError('It is not possible to have negative amounts of pills')
+        self.current_amount = current_amount
+
+    def serialize(self):
+        return {'name': self.name, 'dose': self.dose, 'daily_dose': self.daily_dose, 'current_amount': self.current_amount}
+
```

Comment:
`serialize` zastępuje settery? W tej chwili nie ma za bardzo (choć pewnie jest, ale nie widzę) ochrony przed problemami. Chyba, że to są te cuda, które poczyniłeś w `__init__`.  Ogólnie to strasznie duże te zmiany, nie wiem czy nie zrobię tak, że będę patrzeć sobie tutaj na tego pull requesta robiąc dalej program, ale żebym sam to zrobił, bo teraz masy rzeczy nie ogarniam.

A nazwę `medicine_classes` zmieniłeś na `medicine` dla wygody?

Reply:
`serialize` zamienia instancję na słownik. Nic więcej, nic mniej. Chodzi o to że json jest mocno upośledzony (specjalnie) więc przyjmuje tylko bardzo proste struktury danych. A to oznacza że nie możesz po prostu instancję klasy mu podać.

jak możesz, to każdą większą klasę daj do innego pliku która się zwie tak samo jak klasa. Nie jest to bynajmniej twarda reguła, ale ułatwia potem szukania

Reply:
1. Ja do JSONa (wcześniej pickle'a) chciałem wrzucać tylko info o tym, że dodałem lek, (nazwa ilość i datę), nie całą instancję
2. Spoko, fajne :)

Reply:
A, w ogóle teraz sesja idzie, więc przez jakiś czas mało będę tu robił, ale postaram się przynajmniej ogarnąć to co jest żeby te zmiany były już ogarnięte. Po prostu nie będę prawdopodobnie dalej kodu rozwijał.

Reply:
Całą instancję warto móc też zapisać. Daję ci to możliwość w każdej chwili zapisania do pliku obecny stan programu. To powiedziawszy, rób jak uważasz, twój program, więc lepiej wiesz co chcesz zapisywać :D Chodzi tu tylko o to że json jest ułomny i trza mu bardzo pomagać, więc będziesz musiał mieć takie funkcjiki które ci tworzą odpowiednią strukturę danych

Reply:
Aha, nie zapytałem. To czemu usunąłeś settery w końcu? Nie ma sensu żeby były? Mój wykładowca by zawału dostał jakby to zobaczył :D

Reply:
no. Bo nic nie wnoszą tutaj, tak naprawdę. One mają sens np. w Javie, gdzie masz prywatne pola i defakto musisz użyawć takie badziewa żeby cokolwiek zmieniać. Mają też sens jak robisz dodatkowe operacje. Ale tutaj to jest proste przypisanie i odczyt. Jeżeli wiesz że masz czyste dane, a wiesz bo już wcześniej sprawdzałeś, to możesz poprostu przypisywać i nara. Jak by się w przyszłości okazało że trza coś bardziej skomplikowane zrobić, to byś wtedy na settery/gettery przeszedł.

settery i gettery, jak ogólnie prywatne pola, są po to żebyś mógł swobodnie zmieniać wnętrza klasy, nie psując cudzy kod, gdzie jako "cudzy kod" trza zrozumieć cokolwiek co to wywołuje. Klasa to jest abstrakcja która powinna w zupełności zamknąć jakiś zbiór funkcjonalności. To jest sedno klasy. I najważniejsza jej funkcja. Dlatego jest tak ważne żeby ta abstrakcja była zamknięta (że nie zależy od innych części systemu) i spójna. Tu nie chodzi o jakąś filozoficzne sprawy, tylko żeby to szło ogarnąć głową. Zamykasz jakąś funkcjonalność w klasie i nie przejmujesz się już tym, bo jak ma spójny i zamknięty interfejs, to jak wywołasz jakąś metodę, np. `jedz_do_przodu()`, to nie obchodzi cię jak to robi ani co dokładnie się dzieje - ty wiesz że pojedzie do przodu i tyle.

gettery i settery są cześcią tej umowy jaką jest API klasy. Twórca klasy (w tym ty), daje ci zbiór metod i pól, i mówi coś w stylu "korzystaj tylko z tych metod i pól a będzie dobrze. Obiecuję że one się nie zmienią i będą dalej robiły to co ci teraz mówię. Wnętrze się może bardzo pozmieniać, ale te metody i pola będą ciągle robiły to co ci teraz obiecuję".  Co do prywatnych pól, metod itp., nie masz żadnej obietnicy. Funkcja która się np. zwie `_zmien_na_czerwony()` może dzisiaj pokolorować czerowno, ale jutro na zielono. Albo być usunięta. Nie wiesz. Ale publiczne pola i metody będą cały czas, i będą zawsze robiły to co twierdziły od początku. Przynajmniej takie jest założenie. Tu cały czas chodzi o tą abstrakcję, żeby tobie się łatwiej myślało o systemie. Nie zastanawiasz się ile bezyny na sekundę musisz do silnika wlewać, z jaką mieszanką powietrza, jak często odpalać świece zapłonowe itp., tylko naciskasz pedał gazu. Silnik może zostać cały wymienionwy, ale ty dalej będziesz korzystał tylko z pedałów, biegów i kierownicy. Kierownica może być czysto mechaniczna, albo może być tam miliard elektroniki - ciebie to nie obchodzi, bo korzystasz z abstrakcji jaką jest kierownica, biegi i pedały. Co więcej, każdy samochód korzysta z tej samej abstracji (powiedzmy). To wszystko po to żebyś nie musiał o tym myśleć.

W pythonie, dla użytkonika klasy nie ma żadnej różnicy między:
```
class BezSettery:
    def __init__(self, a, b):
        self.a = a
        self.b = b

class ZSetterami:
    def __init__(self, a, b):
        self._a = a
        self._b = b

    @property
    def a(self):
        return self.a

    @a.setter
    def a(self, value):
       self._a = value

    @property
    def b(self):
        return self.b

    @b.setter
    def b(self, value):
       self._b = value
```
w obu przypadkach tak samo będziesz korzystał z tych klas. Więc tu gettery i settery tylko dokładają więcej kodu bez potrzeby. A jak by się okazało że `a` musi zawsze być jakoś przekształcony przed zapisem, to wtedy dajesz settera. Wcześniej bez sensu.

To wszystko powiedziawszy, jak ktoś nie wie po co to jest, to lepiej mieć te settery i gettery, bo łatwiej potem będzie zmieniać rzeczy. Bo tu znowu chodzi o tą abstrakcję. Jak by ktoś się wcale tym nie przejmował i wszystko pokazywał na zewnątrz, nawet tymczasowe funkcjiki które jutro znikną, to z miesiąc będzie duża kicha jak będzie coś chciał zmienić. Bo się okaże że wszystko jest z sobą poprzeplatane i nic nie da się ruszyć bez psucia wszystkiego. Po to ta abstrakcja. Jak wszystko jest w jednej klasie zamknięte i ma przemyślane fukcje do interakcji, to nawet jak wnętrze klasy to jedna wielka kupa, to przynajmniej reszta programu nie musi się tym przejmować

Reply:
Niech żyje kod Spaghetti

A tak serio, skąplikowany wywód. Rozumiem, że jak nie ogarniam to używać setterów, ale skoro już oragniam to używać jak to potrzebne. Dobrze zrozumiałem? 

Faceta na wykładach pytałem kilka razy na kij to i w sumie to wiele nie powiedział. Za to mieliśmy odruchowo przy inicjowaniu atrybutu w `init` używać `@property` i `@x.setter` xd. W sumie to dlatego nawaliłem tych setterów

Reply:
Pewnie sam nie wiedział... 
Tak se myślę że to nie jest skomplikowane, ale jest trudne :D Że to jest jedno z tych rzeczy które są oczywiste jak wiesz o co chodzi, ale zanim osiągniesz oświecenia to nie masz pojęcia o co chodzi. Trochę jak to że cały zakon można streścić do "Miłuj Boga swego i bliżniego swego", ale to jest oczywiste tylko jak to już rozumiesz, a przed tym to potrzebujesz całe mnóstwo nazkazów i zakazów żeby nie wypaść w jakieś bagno :D

Jak piszesz w pythonie to nie używaj setterów dopóki nie potrzebujesz ich. Jak piszesz w Javie, to używaj setterów chyba że wiesz że nie potrzebujesz ich. To nie są twarde zasady, ale jakoś tak wyszło że tak się robi.

Sprowadza się to do abstrakcji i prostoty (nie łatwości!). Że dobry kod to jest taki który uzywa najprostsze abstrakcje możliwe, ale nie prościej. Gdzie tu prostota jest w sensie brzytwy Ockhama, nie łatwości użycia, bo to są 2 całkiem różne rzeczy. Powtarzam się, ale dlatego że to jest kluczowe - chodzi o to żeby zrobić abstrakcja która całkowicie opisuje to co cię interesuje najmniejszą możliwą liczbą rzeczy. Możesz opisać samolot na wiele różne sposoby, ale przeważnie interesuje cię tylko to że to jest "samolot" i nara. Chcesz wiedzieć czy wejdziesz, czy odleci i czy wyląduje. Reszta cię nie interesuje. Więc twój model samolotu się sprowadza do tych 3 czynników. Samolot to tak naprawdę interacje olbrzymej liczby kwantowych pól. Ale ciebie to totalnie nie obchodzi - chcesz się dostać gdzieś, nie rozważać fizykę. Chyba że chcesz fizykę rozważać, ale wtedy będziesz stosował inny model. 

Nie wiem na ile to jest pomocne... :D Już bardziej rozumiem czemu tak wielu głębokich mądrości brzmi tak banalnie i jest przedstawiono mystycznie albo przypowieściami :D

Reply:
XD Spoko, fajne porównanie. Usunę settery na razie, a co do Javy - ponoć wysysa duszę